manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,367 @@
+
+	// The start of master port range which will hold all the allocation state
+	// of ports allocated so far regardless of whether it was user defined or
+	// not.
+	masterPortStart uint32 = 1
+
+	// The end of master port range which will hold all the allocation state of
+	// ports allocated so far regardless of whether it was user defined or not.
+	masterPortEnd uint32 = 65535
+)
+
+// ErrPortInUse is an error type returned if one of the requested ports is
+// already in use by something else.
+type ErrPortInUse struct {
+	port
+}
+
+// Error returns a formatted message explaining what published port is in use
+func (e ErrPortInUse) Error() string {
+	return fmt.Sprintf("port %d/%s is already reserved", e.port.port, e.port.protocol)
+}
+
+// IsErrPortInUse returns true if the error is of type ErrPortInUse
+func IsErrPortInUse(e error) bool {
+	_, ok := e.(ErrPortInUse)
+	return ok
+}
+
+// PortAllocator is an allocator component that manages the state of which
+// ports (and which protocols) are in use in the the cluster.
+//
+// The PortAllocator totally owns the Endpoint.Ports field. Nothing should be
+// added to Endpoint.Ports outside of the port allocator. If this rule is
+// followed, the the port allocator doesn't have to check for cases like two
+// services having the same port allocated; we created all of the fields, and
+// we can know they are consistent.
+type PortAllocator struct {
+	// ports maps the ports in use. essentially, using map as a set
+	ports map[port]struct{}
+}
+
+// port is the minimal representation for the PortAllocator of a single port.
+// being composed of 2 numeric types, this type is comparable and can be used
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
+	port     uint32
+}
+
+// Proposal is an object that includes the delta result of a call to Allocate
+// or Deallocate, which is passed to Commit to affirmatively alter the ports
+// marked in use
+type Proposal struct {
+	allocate   map[port]struct{}
+	deallocate map[port]struct{}
+}
+
+func NewPortAllocator() *PortAllocator {
+	return &PortAllocator{
+		ports: make(map[port]struct{}),
+	}
+}
+
+// Restore adds the current endpoints to the local state of the port allocator
+// but does not perform any new allocation.
+func (pa *PortAllocator) Restore(endpoints []*api.Endpoint) {
+	if pa == nil {
+		// cannot restore to nil port allocator
+		return
+	}
+	// create and commit a proposal for port allocation
+	proposal := &Proposal{
+		// we only allocate
+		allocate: make(map[port]struct{}),
+	}
+	for _, endpoint := range endpoints {
+		for _, p := range endpoint.Ports {
+			// ignore host-mode ports
+			if p.PublishMode != api.PublishModeHost {
+				proposal.allocate[port{p.Protocol, p.PublishedPort}] = struct{}{}
+			}
+		}
+	}
+	pa.Commit(proposal)
+}
+
+// Commit commits a change to the PortAllocator. There are no safety checks.
+func (pa *PortAllocator) Commit(p *Proposal) {
```

Comment:
I would recommend moving the `Commit` method to the `Proposal` object itself. Checkout the bolt db api for an example of how this would work.

Can this method have an error?

Reply:
I would like to avoid this method having an error. I'd like to avoid having this method. It's literally just because there's no easy way to roll back to a previous PortAllocator state if you abandon allocation.